front/app/utils/streams.ts

```diff
@@ -405,52 +501,71 @@ class Streams {
+        // scan -> The RxJS documentation can probably explain it better than I can: https://www.learnrxjs.io/learn-rxjs/operators/transformation/scan
         startWith('initial' as any),
         scan((accumulated: T, current: T | pureFn<T>) => {
           let data: any = accumulated;
           const dataIds = {};
 
           this.streams[streamId].type = 'unknown';
 
+          // I don't think we still have uses cases were current is a function instead of a value (was an early experiment)
+          // so I wouldn't worry too much about this line. It's safe to assume current will always be an object or an array, as opposed to a function.
+          // If you're feeling brave you could try removing it and try data = current instead :).
           data = isFunction(current) ? current(data) : current;
 
           if (isObject(data) && !isEmpty(data)) {
             const innerData = data['data'];
 
+            // endpoints that return an array of objects
             if (isArray(innerData)) {
               this.streams[streamId].type = 'arrayOfObjects';
+              // loop through the array of objects
               innerData
                 .filter((item) => has(item, 'id'))
                 .forEach((item) => {
                   const dataId = item.id;
                   dataIds[dataId] = true;
                   if (cacheStream) {
+                    // write the endpoint response to the key-value object cache
+                    // first deepfreeze it to guarantee immutability
                     this.resourcesByDataId[dataId] = this.deepFreeze({
                       data: item,
                     });
                   }
+                  // create an index for the stream by its dataId
                   this.addStreamIdByDataIdIndex(
                     streamId,
                     isQueryStream,
                     dataId
                   );
                 });
-            } else if (isObject(innerData) && has(innerData, 'id')) {
+            }
+            // endpoints that return a single object
+            else if (isObject(innerData) && has(innerData, 'id')) {
               const dataId = innerData['id'];
               this.streams[streamId].type = 'singleObject';
               dataIds[dataId] = true;
               if (cacheStream) {
+                // write the endpoint response to the key-value object cache
+                // first deepfreeze it to guarantee immutability
                 this.resourcesByDataId[dataId] = this.deepFreeze({
                   data: innerData,
                 });
               }
+              // create an index for the stream by its dataId
               this.addStreamIdByDataIdIndex(streamId, isQueryStream, dataId);
             }
 
+            // Important: also loop through any included data and put in it the key-value object cache
             if (has(data, 'included')) {
```

Comment:
Would be cool to push the included values to the right streams directly, but I guess it's not so simple either...